go/internal/cmd/copy/copy.go

```diff
@@ -0,0 +1,392 @@
+func parseCommand(line []byte) (*command, error) {
+	s := bytes.TrimSpace(line)
+	after, found := bytes.CutPrefix(s, prefix)
+	if !found {
+		return nil, nil
+	}
+	fields := strings.Fields(string(after))
+	if len(fields) == 0 {
+		return nil, errors.New("empty command")
+	}
+
+	checkArgs := func(want int) error {
+		if got := len(fields) - 1; got != want {
+			return fmt.Errorf("command %q should have %d args, not %d", after, want, got)
+		}
+		return nil
+	}
+
+	d := &command{name: fields[0]}
+	switch fields[0] {
+	default:
+		return nil, fmt.Errorf("unknown command %q", fields[0])
+	case "start":
+		if err := checkArgs(2); err != nil {
+			return nil, err
+		}
+		d.file = fields[1]
+		d.sink = fields[2]
+	case "stop":
+		if err := checkArgs(0); err != nil {
+			return nil, err
+		}
+	case "sink":
+		// sink may have "from src1, src2, ..." after its name.
+		if len(fields)-1 < 1 {
+			return nil, fmt.Errorf("command %q should have at least one arg", after)
+		}
+		d.sink = fields[1]
+	case "endsink":
+		if err := checkArgs(1); err != nil {
+			return nil, err
+		}
+		d.sink = fields[1]
+	}
+	return d, nil
+}
+
+// writeChunks writes the chunks to the destination files.
+func writeChunks(chunks []*chunk) error {
+	// Collect chunks by destination file.
+	byFile := map[string][]*chunk{}
+	for _, c := range chunks {
+		byFile[c.destFile] = append(byFile[c.destFile], c)
+	}
+
+	for file, cs := range byFile {
+		if err := writeChunksToFile(file, cs); err != nil {
+			return fmt.Errorf("%s: %w", file, err)
+		}
+	}
+	return nil
+}
+
+func writeChunksToFile(file string, chunks []*chunk) (err error) {
+	// Parse the destination file into pieces.
+	pieces, err := parseDestFile(file)
+	if err != nil {
+		return err
+	}
+	if err := insertChunksIntoPieces(pieces, chunks); err != nil {
+		return err
+	}
+	data := concatPieces(pieces)
+	return os.WriteFile(file, data, 0644)
+}
+
+// A piece is a contiguous section of a destination file.
+// It is either literal data (sink == "") or a named sink.
+type piece struct {
+	srcFiles map[string]bool
+	sink     string
+	data     []byte
+}
+
+// parseDestFile parses a destination file into pieces.
+func parseDestFile(file string) ([]*piece, error) {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return parseDest(data)
+}
+
+// parseDest parses the contents of a destination file into pieces.
+func parseDest(data []byte) ([]*piece, error) {
+	var pieces []*piece
+	lines := bytes.SplitAfter(data, []byte("\n"))
+	cur := &piece{}
+	for ln, line := range lines {
+		cmd, err := parseCommand(line)
+		if err != nil {
+			return nil, fmt.Errorf("%d: %w", ln, err)
+		}
+		if cmd == nil {
+			cur.data = append(cur.data, line...)
+			continue
+		}
+		switch cmd.name {
+		case "sink":
+			// If the current piece is a sink, then we know now that it is a new sink
+			// and its contents is empty; everything we've seen since the previous sink command
+			// is part of a literal piece.
```

Comment:
That does make it harder on the tool, but easier on people.
Otherwise everywhere you wanted a sink, you'd have to write
```
//copy:sink foo
//copy:endsink foo
```
which is needless busywork: you'll never put anything between those two lines.